### PR TITLE
[WIP] Feature/vm wizard

### PR DESCRIPTION
**Pull request by iceteabottle**
- [x] Add selection for disk offering
- [ ] Format Info card for vm config summary
- [x] Add iso selection
- [x] Add tags to the templates (featured, public, etc.)
- [x] Sort templates/isos
- [x] Pin featured templates/isos at the top of each list
- [x] Add search filter input for isos/templates
- [x] Handle form change on switching from template to iso (and vice versa)
- [ ] Add translations for all input placeholders
- [ ] Implement submit method
- [x]  Add affinity group selection
- [ ] Add network selection
- [x] Add ssh key pairs selection
- [ ] Handle Navigation (e.g. reset the form on navigating away)
- [ ] Add filter possibility to selections (compute offering, disk offering, network offering, ...)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -42,8 +42,7 @@
     "axios": "^0.19.0",
     "enquire.js": "^2.1.6",
     "js-cookie": "^2.2.1",
-    "lodash.get": "^4.4.2",
-    "lodash.pick": "^4.4.0",
+    "lodash": "^4.17.15",
```
Comment (rhtyd) [inline package.json]: +1

File: src/locales/en.json [hunk 2]
```diff
@@ -943,5 +943,6 @@
 "yourInstance": "Your instance",
 "newInstance": "New instance",
 "cpu": "CPU",
-"ram": "RAM"
+"ram": "RAM",
```
Comment (rhtyd) [inline src/locales/en.json]: Can you maybe check against the old UI, wrt any translations you need in the VM deployment form?

**Comment by rhtyd** [comment]
@iceteabottle thanks for resuming this. One request from last time I tested your component/form, can you add the device mixin to hide the infocard on mobile?

**Comment by iceteabottle** [comment]
Current status:

localhost_5050_ (3)

**Comment by iceteabottle** [comment]
Iso Selection (WIP):

Bildschirmfoto 2019-10-30 um 14 41 37

**Comment by rhtyd** [comment]
Nice @iceteabottle - do you think instead we could have an explicit tab of featured templates (use the star icon), rest looks great!

**Comment by iceteabottle** [comment]
> Nice @iceteabottle - do you think instead we could have an explicit tab of featured templates (use the star icon), rest looks great!

Thanks. Maybe I could pin the featured above instead of breaking the OS tabs logic. What do you think, @rhtyd ?

**Comment by rhtyd** [comment]
@iceteabottle that's good idea as well, so basically show features templates? However, per the old CloudStack templates have a sort order so I don't know if that would break anything esp. in env with large number of templates. cc @PaulAngus @svenvogel ?

**Comment by svenvogel** [comment]
@rhtyd Hi Rohit 

in the old wizard there are these tabs.  do you mean this one?
image

or

the order in the templates menu?
image

**Comment by rhtyd** [comment]
Hi @svenvogel in the second screenshot, the order is used to determine the sort order of a resource. CloudStack stores the order in `sort_key` in the `data_center`, `vm_template` and offerings related tables in the cloud db. This sort order is then used to render the sequence of items in the list API response by default, so whatever sort order the admin configures; the templates show up in the UI with that order for each tab (featured, community etc...)

**Comment by svenvogel** [comment]
@rohit ah okay i understand. next question.

if i change the sort order in the second screenshot above it seems it will not be reflected in the old ui wizard. is that possible broken?

**Comment by iceteabottle** [comment]
I re-thought our search solution and added a custom search view instead of filtering over all tabs. I think this is more intuitiv.

Default tab view.
tab-view

Search/filter view:
filter-view

@rhtyd What do you think?

**Comment by rhtyd** [comment]
Good to see you back commiting @iceteabottle !

**Comment by iceteabottle** [comment]
Add affinity group selection.

Bildschirmfoto 2019-12-27 um 14 23 41

File: src/components/view/InfoCard.vue [hunk 3]
```diff
@@ -302,10 +302,19 @@
...
           <a-icon type="calendar" />{{ resource.created }}
         </div>
+        <div class="resource-detail-item" v-if="resource.affinitygroup && resource.affinitygroup.length > 0">
+          <a-icon type="swap" />
+          <span
+            v-for="(group, index) in resource.affinitygroup"
+            :key="group.id"
+          >
+            <router-link :to="{ path: '/affinitygroup/' + group.id }">{{ group.name }}</router-link>
+            <span v-if="index + 1 < resource.affinitygroup.length">, </span>
```
Comment (iceteabottle) [inline src/components/view/InfoCard.vue]: I think that's not the smartest way to display the list of affinity groups. Maybe we can change this later. I would prefer to focus on the wizard and adjust the info card later. Or do you have a suggestion, @rhtyd ?

Bildschirmfoto 2019-12-27 um 14 30 25

**Comment by rhtyd** [comment]
Ping @iceteabottle @svenvogel - any update on this?

**Comment by iceteabottle** [comment]
@rhtyd I'm currently implementing the network selection. That's a complex part.
Did you see my comment here (https://github.com/apache/cloudstack-primate/pull/7#discussion_r361663839)?

**Comment by iceteabottle** [comment]
@rhtyd current status (just a small first part of the network selection):

Bildschirmfoto 2020-01-15 um 18 24 09

**Comment by rhtyd** [comment]
@iceteabottle - let me know if you need any assistance, kindly also merge latest origin/master. Thanks.

**Comment by iceteabottle** [comment]
@rhtyd I wrapped each inner wizard components into a collapsable item to improve the UX.
I will prepare my branch for @utchoang these days.

Bildschirmfoto 2020-01-27 um 16 03 35

**Comment by rhtyd** [comment]
@utchoang please pick this one next, in consultation with @iceteabottle and @svenvogel. Let me know if you've any questions or need any help.

**Comment by iceteabottle** [comment]
@rhtyd Today I will add the last major form part "ssh key pairs selection" to finish the main layout stuff. Can you please merge this PR tomorrow? Then @utchoang can continue my work in his own fork.

**Comment by iceteabottle** [comment]
Add SSH keypair selection.

Bildschirmfoto 2020-01-30 um 15 30 03

**Comment by svenvogel** [comment]
@utchoang can you resume here please?

**Comment by iceteabottle** [comment]
> @utchoang can you resume here please?

@utchoang please see my checklist https://github.com/apache/cloudstack-primate/pull/7#issue-333054299

@rhtyd please merge this PR so @utchoang can finish this feature.

Thanks for the cooperation in this project. I'm sorry that I have to discontinue my work here.

**Comment by utchoang** [comment]
@iceteabottle thank you so much. 
@svenvogel I will start working on it after @rhtyd finishes merging this PR

**Comment by svenvogel** [comment]
@rhtyd can you merge that we can proceed?

**Comment by rhtyd** [comment]
@utchoang @svenvogel I'll squash merge this per your request, I've not tested it. Just checked and added license header to some files.

**Comment by rhtyd** [comment]
@utchoang you can continue on it now, raise a new PR. When you do checkout latest master, ensure it's same as origin master by doing:
```
git fetch origin master
git checkout master
git reset --hard origin/master
git checkout -b 'new-branch-name for your new PR'
```

**Pull request merged**